splink/input_column.py

```diff
@@ -1,112 +1,153 @@
 from __future__ import annotations
 
 from copy import deepcopy
+from dataclasses import dataclass, replace
 
 import sqlglot
 import sqlglot.expressions as exp
-from sqlglot.errors import ParseError
+from sqlglot.errors import ParseError, TokenError
 from sqlglot.expressions import Expression
 
 from .default_from_jsonschema import default_value_from_schema
 
 
-def sqlglot_tree_signature(tree):
-    """
-    A short string representation of a SQLglot tree.
+@dataclass
+class ColumnTreeBuilder:
+    """A builder class that allows you to copy and modify an input column.
 
-    Allows you to easily check that a tree contains certain nodes
+    Modifications copy the base column and can be chained to produce a fresh
+    version of the column.
 
-    For instance, the string "robin['hi']" becomes:
-    'bracket column literal identifier'
+    Columns can also be wrapped in SQLglot expressions by appending them to the
+    base class.
     """
-    return " ".join(n[0].key for n in tree.walk())
 
+    name: str
+    table: str = None
+    quoted: bool = True
+    column_reference: exp.Identifier = None  # the JSON index in 'surname['lat']'
+
+    def unquote(self):
+        return replace(self, quoted=False)
+
+    def exclude_expressions(self):
+        return replace(self, return_column_with_expressions=False)
+
+    def change_name(self, new_name: str):
+        return replace(self, name=new_name)
 
-def add_suffix(tree, suffix) -> Expression:
-    tree = tree.copy()
-    identifier_string = tree.find(exp.Identifier).this
-    identifier_string = f"{identifier_string}{suffix}"
-    tree.find(exp.Identifier).args["this"] = identifier_string
-    return tree
+    def add_prefix(self, prefix: str):
+        return replace(self, name=prefix + self.name) if prefix else self
 
+    def add_suffix(self, suffix: str):
+        return replace(self, name=self.name + suffix) if suffix else self
 
-def add_prefix(tree, prefix) -> Expression:
-    tree = tree.copy()
-    identifier_string = tree.find(exp.Identifier).this
-    identifier_string = f"{prefix}{identifier_string}"
-    tree.find(exp.Identifier).args["this"] = identifier_string
-    return tree
+    def add_table(self, table: str):
+        return replace(self, table=table) if table else self
 
+    def add_alias(self, alias: str) -> exp.Alias:
```

Comment:
It was very unexpected that this returns a sqlglot expression tree - i only noticed this after getting confused over the code in `_table_name_as`

Reply:
Are you suggesting that it makes more sense for this to return a string?

My original view was that the `ColumnTreeBuilder` class would:
* Only return itself (in the case of manipulations) and sqlglot trees
* Handle any manipulations you might want to make to it

Hence why `.add_alias` manipulates the column directly (because we are effectively manipulating the column) and then returns a sqlglot tree.

I believe we have three options from this:
1. Leave the syntax as is.
2. Change the method to take in a sqlglot dialect and return a string.
3. Migrate this across to the `InputColumn` class - which I don't think we should do as it muddies the water too much.

What's your perspective?

Also... on `_table_name_as` - I think the intermediate variables could be better named, so I'll change those now.

Reply:
Oh, and the other thing to note here is that we can change `add_alias` to take in an sqlglot identifier as an argument, which may make the code easier to read.